ballerina/query_complexity_validator_visitor.bal

```diff
@@ -0,0 +1,119 @@
+        self.engine = engine;
+        self.schema = schema;
+        self.maxComplexity = queryComplexityConfig.maxComplexity;
+        self.defaultFieldComplexity = queryComplexityConfig.defaultFieldComplexity;
+        self.warnOnly = queryComplexityConfig.warnOnly;
+        self.operationName = operationName;
+        self.nodeModifierContext = nodeModifierContext;
+    }
+
+    public isolated function visitDocument(parser:DocumentNode documentNode, anydata data = ()) {
+    }
+
+    public isolated function visitOperation(parser:OperationNode operationNode, anydata data = ()) {
+        if self.maxComplexity == 0 {
+            return;
+        }
+        __Type parentType;
+        if operationNode.getKind() == parser:OPERATION_QUERY {
+            parentType = self.schema.queryType;
+        } else if operationNode.getKind() == parser:OPERATION_MUTATION {
+            parentType = <__Type>self.schema.mutationType;
+        } else {
+            parentType = <__Type>self.schema.subscriptionType;
+        }
+        foreach parser:SelectionNode selection in operationNode.getSelections() {
+            selection.accept(self, parentType);
+        }
+        if self.maxComplexity < self.queryComplexity {
+            string operationName = self.operationName == parser:ANONYMOUS_OPERATION ? "" : string `${self.operationName} `;
+            string message = string `The operation ${operationName}exceeds the maximum query complexity threshold. Maximum allowed complexity: ${self.maxComplexity}, actual complexity: ${self.queryComplexity}`;
+            if self.warnOnly {
+                log:printWarn(message);
+            } else {
+                self.errors.push(getErrorDetailRecord(message, operationNode.getLocation()));
+            }
+        }
+    }
+
+    public isolated function visitField(parser:FieldNode fieldNode, anydata data = ()) {
+        __Type? 'type = <__Type?>data;
+        __Type fieldType;
+        if 'type is __Type {
+            __Type parentType = getOfType('type);
+            string coordinate = string `${parentType.name.toString()}.${fieldNode.getName()}`;
+            int|Error fieldComplexity = getFieldComplexity(self.engine, coordinate);
+            if fieldComplexity is Error {
+                log:printDebug(string `Complexity not found for field: "coordinate"`, fieldComplexity);
+            }
+            int complexity = fieldComplexity is int ? fieldComplexity : self.defaultFieldComplexity;
+            self.queryComplexity += complexity;
+            __Field? requiredFieldValue = getRequierdFieldFromType(parentType, self.schema.types, fieldNode);
+            if requiredFieldValue is () {
+                return;
+            }
+            fieldType = requiredFieldValue.'type;
+            foreach parser:SelectionNode selection in fieldNode.getSelections() {
+                selection.accept(self, fieldType);
+            }
+        }
+        // TODO: Handle introspection queries
```

Comment:
Did we create a separate issue to track this?

Reply:
We can handle this scenario. I will update the code.

Reply:
Updated the code and the spec.